secp256k1-sys/src/recovery.rs

```diff
@@ -72,58 +74,77 @@ extern "C" {
-                                                   msg32: *const c_uchar,
-                                                   sk: *const c_uchar,
-                                                   _noncefn: NonceFn,
-                                                   _noncedata: *const c_void)
-                                                   -> c_int {
-        assert!(!cx.is_null() && (*cx).flags() & !(SECP256K1_START_NONE | SECP256K1_START_VERIFY | SECP256K1_START_SIGN) == 0);
-        assert!((*cx).flags() & SECP256K1_START_SIGN == SECP256K1_START_SIGN);
-        if secp256k1_ec_seckey_verify(cx, sk) != 1 { return 0; }
-        if *sk.offset(0) > 0x7f {
-            (*sig).0[0] = 2;
-        } else {
-            (*sig).0[0] = 3;
+    use super::*;
+    use std::slice;
+
+    use secp256k1_ec_pubkey_create;
+    use secp256k1_ec_pubkey_parse;
+    use secp256k1_ec_pubkey_serialize;
+    use SECP256K1_SER_COMPRESSED;
+
+    /// Sets sig to msg32||full pk
+    pub unsafe fn secp256k1_ecdsa_sign_recoverable(
+        cx: *const Context,
+        sig: *mut RecoverableSignature,
+        msg32: *const c_uchar,
+        sk: *const c_uchar,
+        _noncefn: NonceFn,
+        _noncedata: *const c_void,
+    ) -> c_int {
+        // Check context is built for signing (and compute pk)
+        let mut new_pk = PublicKey::new();
+        if secp256k1_ec_pubkey_create(cx, &mut new_pk, sk) != 1 {
+            return 0;
         }
-        ptr::copy(msg32, (*sig).0[1..33].as_mut_ptr(), 32);
-        ptr::copy(sk, (*sig).0[33..65].as_mut_ptr(), 32);
+        // Sign
+        let sig_sl = slice::from_raw_parts_mut(sig as *mut u8, 65);
+        let msg_sl = slice::from_raw_parts(msg32 as *const u8, 32);
+        sig_sl[..32].copy_from_slice(msg_sl);
+        let mut out_len: size_t = 33;
+        secp256k1_ec_pubkey_serialize(cx, sig_sl[32..].as_mut_ptr(), &mut out_len, &new_pk, SECP256K1_SER_COMPRESSED);
+        // Encode the parity of the pubkey in the final byte as 0/1,
+        // which is the same encoding (though the parity is computed
+        // differently) as real recoverable signatures.
+        sig_sl.swap(32, 64);
+        sig_sl[64] -= 2;
         1
     }
 
-    pub unsafe fn secp256k1_ecdsa_recover(_cx: *const Context,
-                                          _pk: *mut PublicKey,
-                                          _sig: *const RecoverableSignature,
-                                          _msg32: *const c_uchar)
-                                          -> c_int {
-        unimplemented!();
+    pub unsafe fn secp256k1_ecdsa_recover(
+        cx: *const Context,
+        pk: *mut PublicKey,
+        sig: *const RecoverableSignature,
+        msg32: *const c_uchar
+    ) -> c_int {
+        let sig_sl = slice::from_raw_parts(sig as *const u8, 65);
+        let msg_sl = slice::from_raw_parts(msg32 as *const u8, 32);
+            println!("HMM0");
+
+        if sig_sl[64] > 4 {
```

Comment:
I think that should be `>= 4`, because the VERIFY_CHECK makes sure that `recid < 4`, so it should return 0 on 4.